### PR TITLE
TTO-276 Revise queueing order in CRMS

### PR DESCRIPTION
**Pull request by moseshll**
TL;DR for reviewer: a quick scan suffices, read on if fascinated with minutiae.

Note to reviewer: this has been tested on dev-2. This is a somewhat inconsequential tweak to the runtime behavior of the system. Basically we want to promote "mdp" volumes to the front of the queue and the front of the candidates.  `PresentationOrder` is an existing mechanism (note camel case) and `queue_order` is the new method (note snake case) using almost the same mechanism, just applied to a different query. (The mechanism is embarrassingly clunky but I don't know how else to do it: basically we allow a Project to respond with its own `ORDER BY` clause that gets plugged into the caller's query.)

I have an issue (#195) to change the return values of these subclass methods. Out of scope for this ticket.

- Add a custom `PresentationOrder` to the Core project to prioritize mdp namespace.
- Add a new `Project#queue_order` used the same way as `PresentationOrder` except:
  - It applies to the list of candidates being considered for addition to the Queue.
  - Like `PresentationOrder` the custom `ORDER BY` becomes the first order clause.
  - And the default behavior is to return `undef`.

### CODE REVIEW

File: cgi/Project.pm [hunk 1]
```diff
@@ -10,9 +10,11 @@ sub new
   $self->{$_} = $args{$_} for keys %args;
   my $id = $args{'id'};
   die "No CRMS object passed to project" unless $args{'crms'};
-  my $sql = 'SELECT * FROM projects WHERE id=?';
-  my $ref = $self->{'crms'}->get('dbh')->selectall_hashref($sql, 'id', undef, $id);
-  $self->{$_} = $ref->{$id}->{$_} for keys %{$ref->{$id}};
+  if (defined $id) {
```
Comment (moseshll) [inline cgi/Project.pm:13]: I added this so we could create a Project instance without an id (primary key). Normally we never instantiate the Project class, just the subclasses, but this gives me a chance to test the default behavior without writing tests for all of the subclasses (definitely out of scope for this ticket).

File: cgi/CRMS.pm [hunk 2]
```diff
@@ -1458,7 +1462,7 @@ sub LoadQueueForProject {
     $sql = 'SELECT c.id FROM candidates c'.
            ' INNER JOIN bibdata b ON c.id=b.id'.
            ' WHERE c.project=? AND b.sysid=?'.
-           ' ORDER BY c.time ASC';
```
Comment (moseshll) [inline cgi/CRMS.pm:1461]: `c.time ASC`? `c.time DESC`? The second query is for a subset of the first and the order doesn't really matter so we'll just reuse the `ORDER BY` from the outer query.

Comment (mwarin) [reply cgi/CRMS.pm:1461]: If the order "doesn't really matter" here, why order at all?

File: cgi/Project.pm [hunk 3]
```diff
@@ -93,6 +95,16 @@ sub EvaluateCandidacy
 }
 
 # ========== REVIEW INTERFACE ========== #
+# Called by CRMS::LoadQueueForProject to prioritize candidates for the queue.
+# Return undef for no additional order (the default), or
+# a column name in bibdata (b.*) or candidates (c.*).
+# Example: 'b.author DESC'
+sub queue_order {
```
Comment (mwarin) [inline cgi/Project.pm]: If the body of this sub is essentially a no-op, I'd try just `sub queue_order {}`. I _think_ that's equivalent.

File: cgi/CRMS.pm [hunk 4]
```diff
@@ -1436,19 +1436,23 @@ sub LoadQueueForProject {
...
   my $sql = 'SELECT COUNT(*) FROM queue WHERE project=?';
   my $queueSize = $self->SimpleSqlGet($sql, $project);
-  my $targetQueueSize = $self->GetProjectRef($project)->queue_size();
+  my $project_ref = $self->GetProjectRef($project);
+  my $targetQueueSize = $project_ref->queue_size();
   my $needed = $targetQueueSize - $queueSize;
   $self->ReportMsg("Project $project_name: $queueSize volumes -- need $needed");
   return if $needed <= 0;
   my $count = 0;
+  my @orders = ('c.time DESC');
```
Comment (mwarin) [inline cgi/CRMS.pm:1445]: I would probably have done it like:

```perl
# value(s) for ORDER BY: include & prioritize "project order" if defined
my $orders = join(' ', grep { defined $_ } ($project_ref->queue_order, 'c.time DESC'));
```

... but there is of course more than one way to do it.

**Review by mwarin — APPROVED** [review]
Looks fine, nothing that stands in the way of an APPROVE.

File: cgi/CRMS.pm (hunk 2 above)
Comment (moseshll) [reply cgi/CRMS.pm:1461]: Good point. Removed from second query.

File: cgi/CRMS.pm (hunk 4 above)
Comment (moseshll) [reply cgi/CRMS.pm:1445]: I have a GH issue to redefine these optional methods as returning an arrayref, default empty. That way it's just an array merger without having to deal with `undef`. Out of scope for a TTO, but yeah definitely could do better.

File: cgi/Project.pm (hunk 3 above)
Comment (moseshll) [reply cgi/Project.pm]: Sure is, works for me.
